dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,517 @@
+        """
+        returns an iterator of all nodes that are one step
+        from the current node down including through iterables
+        """
+        return self.fields(flat=True, nodes_only=True, obfuscated=False, nones=False)
+
+    def filter(self, func: Callable[["Node"], bool]) -> Iterator["Node"]:
+        """
+        find all nodes that `func` returns `True` for
+        """
+        if func(self):
+            yield self
+        for node in chain(*[child.filter(func) for child in self.children]):
+            yield node
+
+    def find_all(self, node_type: Type["Node"]) -> Iterator["Node"]:
+        """
+        find all nodes of a particular type in the node's sub-ast
+        """
+        return self.filter(lambda n: isinstance(n, node_type))
+
+    def apply(self, func: Callable[["Node"], None]):
+        """
+        traverse ast and apply func to each Node
+        """
+        func(self)
+        for child in self.children:
+            child.apply(func)
+
+    def compare(self, other: "Node") -> bool:
+        """
+        compare two ASTs
+        """
+
+        return self == other and all(
+            child.compare(other_child)
+            for child, other_child in zip_longest(self.children, other.children)
+        )
+
+    def __eq__(self, other) -> bool:
+        """
+        Compares two nodes for "top level" equality.
+        Checks for type equality and primitive field types for full equality.
+        Compares all others for type equality only. No recursing.
+        Note: Does not check (sub)AST. See `Node.compare` for comparing (sub)ASTs.
+        """
+        primitives = {int, float, str, bool, type(None)}
+        return type(self) == type(other) and all(  # pylint: disable=C0123
+            s == o
+            if type(s) in primitives  # pylint: disable=C0123
+            else type(s) == type(o)  # pylint: disable=C0123
+            for s, o in zip(
+                (self.fields(False, False, False, True)),
+                (other.fields(False, False, False, True)),
+            )
+        )
+
+    @abstractmethod
+    def __hash__(self) -> int:
+        """
+        hash a node
+        """
+
+
+class Expression(Node):
+    """an expression type simply for type checking"""
+
+
+@dataclass(eq=False)  # type: ignore
+class Named(Expression):
+    """An Expression that has a name"""
+
+    name: str
+    quote_style: Optional[str]
+
+    @property
+    def quoted_name(self) -> str:
+        """
+        get the name of the Named Node including the quotes if any
+        """
+        return f'{self.quote_style if self.quote_style else ""}{self.name}{self.quote_style if self.quote_style else ""}'  # pylint: disable=C0301
+
+    def alias_or_name(self) -> str:
+        """
+        get the name or alias of the node
+        """
+        if len(self.parents) == 1:
+            parent = tuple(self.parents)[0]
+            if isinstance(parent, Alias):
+                return parent.name
```

Comment:
Why do we return the parent name when there's a single parent?

Reply:
This is borrowed from sqlglot that it is useful to have a method return the alias name if an alias exists otherwise return the node's name so this should only return the parent name if there is a single alias node as the parent